### ISSUE TITLE
NiceHash feature [Bounty]

### ISSUE THREAD

**Issue by slayerulan** [Snipa22/xmr-node-proxy#89]
Hi,
I will donate 2 XMR to anyone who successfully integrate NiceHash to xmr-node-proxy

Thanks

**Comment by Snipa22** [comment]
Literally impossible.  The work must be done by NiceHash, and past that, the pools that NH connects to.  The issue is that NH is a proxy.  You can't proxy a proxy.

**Comment by windrunner414** [comment]
@Snipa22 why?xmrig-proxy can do it.

**Comment by Snipa22** [comment]
No, it can't.  It can accept NH-style connections, but it can't connect to NH, because it's not compatible.  XNP can accept NH style connections fine.

**Comment by tostercx** [comment]
It does seem to be "possible" tho I'm not sure of the drawbacks:
https://github.com/xmrig/xmrig-proxy/pull/83

**Comment by untoreh** [comment]
@tostercx no it doesn't, xmrig-proxy can change the diffculty only to a higher one, it can't lower the pool difficulty like xmr-node-proxy, which is in general used to prevent to pool from submitting higher diff jobs and ending up with timeouts. With nicehash you would never encounter this problem because the pool will most likely never submit jobs with a difficulty higher than 200k

**Comment by tostercx** [comment]
@untoreh yeah I took a closer look at that PR and it basically does 1:1 connections with NH... the same could be accomplished with any SOCKS proxy, so I'm not sure what the author uses it for.

What would be cool if there was a proxy to allow multiple slow miners sit on a single NH connection.

It should be possible if we further split the nonce into sub-ranges similarly to how NH does it by "locking a byte". That would require patches to the protocol, proxy and miner software tho.

**Comment by oldschoola41** [comment]
@tostercx could you do this?

**Comment by tostercx** [comment]
@oldschoola41 no, I'm too lazy.

**Comment by oldschoola41** [comment]
@tostercx lol what about % of mining after you develop it

**Comment by tostercx** [comment]
@oldschoola41 it's a lot of work and I really don't have time atm.

The proxy would have to dynamically estimate how much hashing power it has and how many connections it needs keep open to NH. Then it would need to estimate every miner's hashing power and allocate it a "range" on a NH connection/nonce it can mine on. I have no idea how hard it is to patch existing proxies, it might need a new implementation.

Patching the protocol should be easy but then on the miner's end you'd need new logic to distribute that range further down for each thread it runs... which also sounds like a PITA. And then you'd probably need some kind of a mechanism to get another range from the proxy in case it's a slow block / you run out of mining space in your old range.

A _lot_ of work... not gonna do it, sorry.

**Comment by oldschoola41** [comment]
@tostercx understood, thx anyway!

**Comment by taherkaraki** [comment]
I know this is a pretty old post, but I only started mining very recently and was introduced to nicehash. So I downloaded the app, saw the command-line that Nicehash uses to run XMRig with NH’s RandomX pool. Then I took the config and amended as needed in the XMRig-proxy. I then linked some machines using --nicehash argument to xmrig-proxy and things were rosy for like 10 days. I was happily receiving jobs and all machines were accepting them normally. Until it stopped. Suddenly from two days ago all jobs are getting rejected with “invalid nonce; not compatible with NiceHash?”. I mean I didn’t even know it was not possible to chain proxies until reading about it here following my issue.

My question is how on earth was it working for me for a mere 10 days?
